### PR TITLE
Reactive MongoDB with Panache

### LINKED ISSUE #5446: Reactive MongoDB with Panache
**Describe the extension**
A reactive version of MongoDB with Panache based on the existing `ReactiveMongoClient` provided by the `quarkus-mongodb-client` extension.

Interested in this extension, please +1 via the emoji/reaction feature of GitHub (top right).

**Configuration suggestion**
The configuration should be the same as the one from MongoDB with Panache.

**Additional context**
The API should be the same than the one from MongoDB with Panache but using the reactive client and returing Reactive Stream types from MP Reactive Stream.

### PR DESCRIPTION
**Pull request by loicmathieu**
A reactive version of the MongoDB with Panache extension.

The implementation needs some polishing and lack documentation, this will be done if the extension is agreed to be merged.

There is some open questions:

- I use the `-rx` suffix sometimes but I think it's wrong as it is not about reactive extension but reactive streams so I should only use `reactive`
- I'm not sure that the implementation of `ReactiveMongoOperations.update(ReactiveMongoCollection collection, List<Object> entities)` is optimal. I use `thenCompose` to linked the future one after the other maybe creating all futures and using CompletableFuture.allOf() will be better ?
- I'm not sure that the implementation of `ReactivePanacheQueyImpl.singleResult()` is ideal. 

I didn't add a test on the `stream()` operations that returns a `Publisher`, this should be handled differently than a return type of `CompletionStage` from Resteasy (at least when producing server side events content type) so maybe I should add one to the test.

Thanks @cescoffier for the reactive mongodb client and @FroMage from the API build on it's prototype on Reactive Hibernate with Panache, without them I will not have been able to do this extension :)

Fixes #5446

### CODE REVIEW

**Comment by FroMage** [comment]
The extension naming question is something I don't know how to answer. Do we have a policy @cescoffier ?

I use `thenCompose` too, rather than `CompletableFuture.allOf` to avoid creating futures. Not sure if that's good or bad, but I do it this way too, especially since `allOf` does not let me get the results.

As for the `Publisher` yes you should test it as well. In RESTEasy the default is to collect them into a `List<T>` so it's easy to test.

**Review by FroMage — CHANGES_REQUESTED** [review]
LGTM, except some namiing/type decisions we have to take.

File: extensions/panache/mongodb-panache-rx/runtime/src/main/java/io/quarkus/mongodb/panache/rx/PanacheMongoEntityBase.java
Comment (loicmathieu) [reply extensions/panache/mongodb-panache-rx/runtime/src/main/java/io/quarkus/mongodb/panache/rx/PanacheMongoEntityBase.java]: I think that we didn't want to offer a `java.util.Stream` for reactive stream because it will confuse the user as they are lazy by natures so you will end up with stream lazyness and reactive stream asynchronous processing that are also lazyness in a some way.
I would keep it like this or change the method name. But definitly not offering a method that return a `java.util.Stream`.

As always, better propose less then too much as we can add methods later but it's difficult to remove them after some user bet used to them :)

File: extensions/panache/mongodb-panache-rx/runtime/src/main/java/io/quarkus/mongodb/panache/rx/PanacheMongoEntity.java
Comment (emmanuelbernard) [reply extensions/panache/mongodb-panache-rx/runtime/src/main/java/io/quarkus/mongodb/panache/rx/PanacheMongoEntity.java]: I’m tempted to say yes. Have a Reactive prefix or Suffix on API contracts you avoid import mistakes.

Comment (loicmathieu) [reply extensions/panache/mongodb-panache-rx/runtime/src/main/java/io/quarkus/mongodb/panache/rx/PanacheMongoEntity.java]: Is it Panache or Mongo that will be reactive ? I would say Panache! So I propose to name it "ReactivePanacheMongo<Stuff>". 

Then I ask myself what will be the "display name" : Reactive MongoDB with Panache? MongoDB with Reactive Panache ?

Namming is hard :)

File: extensions/panache/mongodb-panache-rx/deployment/src/main/java/io/quarkus/mongodb/panache/rx/deployment/PanacheResourceProcessor.java
Comment (loicmathieu) [reply extensions/panache/mongodb-panache-rx/deployment/src/main/java/io/quarkus/mongodb/panache/rx/deployment/PanacheResourceProcessor.java]: Yes, and it is now :)

File: extensions/panache/mongodb-panache-rx/runtime/src/main/java/io/quarkus/mongodb/panache/rx/PanacheMongoEntity.java
Comment (loicmathieu) [reply extensions/panache/mongodb-panache-rx/runtime/src/main/java/io/quarkus/mongodb/panache/rx/PanacheMongoEntity.java]: I rename everything with Reactive

**Comment by loicmathieu** [comment]
@FroMage I test the `Publisher` case with SSE, I find that there is no JSON serialization by default with SSE inside Quarkus so it uses the `toString()` of the entity to serialize the entity :(

I can remove the SSE header so that Resteasy serialize it as a list (already tested and it works). But I'm curious of how to do SSE with JSON payload using Quarkus resteasy extension (on the resteasy site they is not a lot of documentation for this: https://docs.jboss.org/resteasy/docs/3.5.1.Final/userguide/html/JAX-RS_2.1_additions.html#d4e1430

**Comment by FroMage** [comment]
> @FroMage I test the Publisher case with SSE, I find that there is no JSON serialization by default with SSE inside Quarkus so it uses the toString() of the entity to serialize the entity :(

It totally serialises over json. What does your test look like? Perhaps we have a bug.

**Comment by loicmathieu** [comment]
> It totally serialises over json. What does your test look like? Perhaps we have a bug.

Yes, I think so :(
See this commit: https://github.com/quarkusio/quarkus/pull/5567/commits/f25ee3281471057c8d8e825ab1eb66cebcd19546

I return a `Publisher<BookEntity>` with a SSE Header and inside the test I use a restclient to receive those SSE events, I receive 3 events with a text reprensentation (the toString of the BookEntity) and not a JSON representation: https://github.com/quarkusio/quarkus/blob/f25ee3281471057c8d8e825ab1eb66cebcd19546/integration-tests/mongodb-panache-rx/src/test/java/io/quarkus/mongodb/panache/rx/BookResourceTest.java#L222

**Comment by FroMage** [comment]
You're missing a content type for the entity, see `@SseElementType` and https://docs.jboss.org/resteasy/docs/4.4.1.Final/userguide/html/Reactive.html#d4e2483

File: extensions/panache/mongodb-panache-rx/deployment/src/main/java/io/quarkus/mongodb/panache/rx/deployment/PanacheResourceProcessor.java
Comment (Sanne) [reply extensions/panache/mongodb-panache-rx/deployment/src/main/java/io/quarkus/mongodb/panache/rx/deployment/PanacheResourceProcessor.java]: To be fair, we discussed a long time ago that there's no reason to have all those constants collected in the same class.

It's just a source of likely conflicts among us all.

I'm not sending a PR which deletes it, as people are still regularly editing it in PRs, so me doing such a "cleanup" would get in the way for no good reason either...

But you could all definitely stop doing it, I'd just remove the TODO and keep the constant here.

**Comment by loicmathieu** [comment]
> You're missing a content type for the entity, see `@SseElementType` 

Thanks a lot!! I search for this and didn't find it inside the Reasteasy documentation.
We definitly need to include it in our reactive guide, all examples are based on text representation and it's not really what a user would use.

File: extensions/panache/mongodb-panache-rx/deployment/src/main/java/io/quarkus/mongodb/panache/rx/deployment/PanacheResourceProcessor.java
Comment (loicmathieu) [reply extensions/panache/mongodb-panache-rx/deployment/src/main/java/io/quarkus/mongodb/panache/rx/deployment/PanacheResourceProcessor.java]: @Sanne I just move it to `FeatureBuildItem` yesterday :(
I can restore it later if you want

Comment (Sanne) [reply extensions/panache/mongodb-panache-rx/deployment/src/main/java/io/quarkus/mongodb/panache/rx/deployment/PanacheResourceProcessor.java]: no worries, feel free to leave it there :)

I couldn't stop myself from reminding about it buy there' no need for that minor thing to delay this PR.

**Comment by loicmathieu** [comment]
@FroMage I added Optional support like what was done for MongoDB with Panache.

@cescoffier can you give your feedback on the reactive part of the PR ?

**Comment by cescoffier** [comment]
The reactive API looks ok to me. Note that we are considering moving away from CompletionStage.

**Comment by loicmathieu** [comment]
> Note that we are considering moving away from CompletionStage.

@cescoffier from microprofile reactive stream? Away to what ?

**Comment by cescoffier** [comment]
@loicmathieu we are working on a new reactive programming library.

**Comment by emmanuelbernard** [comment]
@loicmathieu The new work is here https://github.com/smallrye/smallrye-mutiny

While before our preferred way was `CompletionStage` and `Publisher` (and MP Reactive Stream Operators), we realized that `CompletionStage` was giving a pretty bad API experience mainly because it was an eager process and bridges to/from CompletionStage were cludgy.

So we looked around for an approach that would be forward thinking (Java 11+), work on GraalVM without massive reflection, generally neutral community or with open governance. But none fit the bill. So we have been exploring what we would do to bring reactive types to the masses and keep bridges to other reactive types open by design. And that's what's in the repo. Clement is giving it a few extra days of work because announcing it.

**Comment by loicmathieu** [comment]
> bring reactive types to the masses and keep bridges to other reactive types open by design

This sounds interesting ;)

Should we keep the current implementation and merge this soon (it's almost ready) or should we wait a few day (and miss the 1.1 target) and refacto to use this new programming library ?

**Comment by emmanuelbernard** [comment]
I'm not sure to be honest, I'm open for feedback. To me it's more a factor of whether the API will end up different than PanacheRX RDBMS CC @FroMage 
But if we wait for PanacheRX RDBMS this one is one ice for a while.

**Comment by loicmathieu** [comment]
The _current_ API of Reactive Hibernate with Panache (PanacheRX RDBMS - but we decided to not use RX inside the name :) ) is the same as this one, I don't know if there are plan to move it to something else.

We can still merge it on preview and keep it in the preview state as long as we didn't ship the Hibernate flavor to be sure that APIs are aligned ...

**Comment by FroMage** [comment]
We can merge this as-is and provide an alternate class hierarchy for Mutiny. Or wait and use Mutiny, but what will our policy be @cescoffier ? Keep on providing Axle APIs in parallel, or axe (haha) them in favour of Mutiny?

**Comment by loicmathieu** [comment]
I prefere not to wait :)

In order to be able to propose multiple reactive APIs in the future, maybe move everything to `io.quarkus.panache.mongodb.reactive.stream` to be able to have a `io.quarkus.panache.mongodb.reactive.mutiny` package later ?

When Mutiny will be integrated inside Quarkus we will need to review some APIs anyway so it will be the time to review this one also :)

**Comment by FroMage** [comment]
Probably `io.quarkus.panache.mongodb.reactive.axle` which is what we call the current system.

**Comment by cescoffier** [comment]
+1 to Stephane's proposal. The current model is named `axle` so let's use this in the name of the package. The future model is named `mutiny`. I plan the first beta release on Friday.

**Comment by emmanuelbernard** [comment]
@FroMage @cescoffier so my argument against just the package name being different is incorrect class import by accident

**Comment by loicmathieu** [comment]
@emmanuelbernard the idea is not to provide both implementations, but to be able to smoothly transition from one to another by deprecating everything inside the `axle` package, when we create the new `mutiny` package.

If we want to support both implementation, having the same classes in the same package will definitly leads to error. In this case I prefere to offers a new extension using mutiny ...

**Comment by FroMage** [comment]
@emmanuelbernard agreed.

BUT `ReactivePanacheMongoEntity` already has three tokens just for damn specifiers. At some point it becomes ridiculous.

In retrospect, I think we should have gone with `MongoPanacheEntity` (to keep completion easy) or even `MongoEntity` (if we didn't care about "branding").

So anyway, if we need to put the flavour in the name, we can only replace `Reactive` with `Axle` or `Mutiny`: I don't think we can afford `ReactiveAxlePanacheMongoEntity` with a straight face :(

**Comment by emmanuelbernard** [comment]
> BUT `ReactivePanacheMongoEntity` already has three tokens just for damn specifiers. At some point it becomes ridiculous.

Technicall it is a `Factory` and a `Manager` pattern, so...

**Comment by loicmathieu** [comment]
My concern is that besides you guys, nobody knows what Axle is ...

So, we can do three different things to smooth up the upcomming transition :

1. Use different extension
2. Use different package
3. Use different name (so rename to `ReactiveAxlePanacheMongoEntity` or decide to remove Mongo or even PanacheMongo) for the name for conciseness

Thinking must be done carefully as what will be done for other extensions that support Reactive Stream (reactive messaging, reactive database, mongo-client, ...).

If we have no plan yet about how Mutiny will be integrated into the other extension, I think using a package qualified by `axle` or `stream` (as it uses a reactive-stream implementation) is OK because we could still use any of the three other solution for the Mutiny implementation later ...

**Comment by emmanuelbernard** [comment]
Probably ReactiveXx is fine for the Axle plan. And MutinyXx for the Mutiny specific one? 
The package too as this one is cheap.
BTW both Axle and Mutiny are Reactive Streams implemenations @loicmathieu

**Comment by emmanuelbernard** [comment]
Probably ReactiveXx is fine for the Axle plan. And MutinyXx for the Mutiny specific one? 
The package too as this one is cheap.
BTW both Axle and Mutiny are Reactive Streams implemenations @loicmathieu

**Comment by loicmathieu** [comment]
@emmanuelbernard as I understand it, mutiny provides reactive types (Multi and Uni) on top of Reactive Streams. 
For Axle, well, I don't know as I directly use CompletionStage and Publisher with Axle wich are standard reactive types :)

File: extensions/panache/mongodb-panache-rx/runtime/src/main/java/io/quarkus/mongodb/panache/rx/PanacheMongoEntityBase.java
Comment (loicmathieu) [reply extensions/panache/mongodb-panache-rx/runtime/src/main/java/io/quarkus/mongodb/panache/rx/PanacheMongoEntityBase.java]: @FroMage do you want me to change the signature or not? @cescoffier didn't give it's idea on it but if we do it for `persist` we should also do it for `persistOrUpdate` and for `update`.

File: extensions/panache/mongodb-panache-rx/runtime/src/main/java/io/quarkus/mongodb/panache/rx/PanacheMongoEntityBase.java
Comment (loicmathieu) [reply extensions/panache/mongodb-panache-rx/runtime/src/main/java/io/quarkus/mongodb/panache/rx/PanacheMongoEntityBase.java]: This is not required but would make all non static method signature easier as they could use the T type parameter from the type instead of defining a type parameter for each method.

As Hibernate and Mongo panache extension didn't add a type type parameter I propose to not provide one here. This can be review later and added to the three extension without breaking compatibility.

**Comment by loicmathieu** [comment]
I move to the `io.quarkus.mongodb.panache.reactive.axle` package and keep the same names as suggested by  @emmanuelbernard .

**Comment by loicmathieu** [comment]
@FroMage is there still some work to do on the implementation side ? I would want to target 1.2 for this and will be happy to merge it before the end of the year :)

@gsmet I need to write a guide and a quickstart for this. Do I create a new guide "Reactive MongoDB with Panache" or should I add a reactive section inside the existing guide ?

**Comment by FroMage** [comment]
> Probably ReactiveXx is fine for the Axle plan. And MutinyXx for the Mutiny specific one?

@cescoffier reactive strategy check here:

- package is `io.quarkus...axle`
- name is `ReactiveXXX`
- mutiny version will be in `io.quarkus...mutiny` under `MutinyXXX` name

Is that our official policy?

**Comment by FroMage** [comment]
So our official policy:

    - blocking and reactive should be in a single module
    - root is the blocking API (io.quarkus.mongo)
    - axle is the axle reactive API (io.quarkus.mongo.axle)
    - mutiny is the mutiny reactive API (io.quarkus.mongo.mutiny)
    - reactive classes are prefixed with `Reactive`.
    - when mutiny lands we deprecate the axle classes, thus 
      making their class names distinguishable from the mutiny version with the same name.

**Comment by loicmathieu** [comment]
@FroMage all reactive stuff moved to the same extension on an axle package.
Integration tests merged into one module.

File: integration-tests/mongodb-panache/src/main/java/io/quarkus/it/mongodb/panache/axle/book/ReactiveBookEntityResource.java
Comment (FroMage) [inline integration-tests/mongodb-panache/src/main/java/io/quarkus/it/mongodb/panache/axle/book/ReactiveBookEntityResource.java]: We should really add a `deleteById(id)` operation, no?

**Review by FroMage — APPROVED** [review]
LGTM

File: integration-tests/mongodb-panache/src/main/java/io/quarkus/it/mongodb/panache/axle/book/ReactiveBookEntityResource.java
Comment (loicmathieu) [reply integration-tests/mongodb-panache/src/main/java/io/quarkus/it/mongodb/panache/axle/book/ReactiveBookEntityResource.java]: I would love to, and in both MongoDB and Hibernate Panache!
I'll open an issue for it so we will see if someone else would love it :)

Comment (loicmathieu) [reply integration-tests/mongodb-panache/src/main/java/io/quarkus/it/mongodb/panache/axle/book/ReactiveBookEntityResource.java]: See the PR here: #6270

**Comment by FroMage** [comment]
Hold on, I don't see docs, is it normal? I thought I had seen them, but perhaps I'm confusing this with another PR?

**Comment by loicmathieu** [comment]
@FroMage no doc yet.
I wil work on it next week.
We can hold on the PR untill the documentation or merge it and provides documentation later (it targets 1.2 that will not be delivered soon)

**Review by FroMage — CHANGES_REQUESTED** [review]
In that case, let me request docs to avoid accidental merges :)

**Comment by loicmathieu** [comment]
@FroMage I added a section in the existing MongoDB with Panache guide

File: docs/src/main/asciidoc/mongodb-panache.adoc
Comment (FroMage) [inline docs/src/main/asciidoc/mongodb-panache.adoc]: Indentation issue?

**Review by FroMage — CHANGES_REQUESTED** [review]
LGTM aside from the indentation issue.

File: docs/src/main/asciidoc/mongodb-panache.adoc
Comment (loicmathieu) [reply docs/src/main/asciidoc/mongodb-panache.adoc]: Fixed!

**Review by FroMage — APPROVED** [review]

**Comment by loicmathieu** [comment]
@gsmet this PR has been approved, I would love to see it included for 1.2 :)

**Comment by dufoli** [comment]
just my 2 cents, 
- does it work with new multi mongo #4529 ?
- I see that you have a fix ever merged, So I guess it is not rebase on master
- I see that you have commit fix, it make no sens to have intermediate commit with issue so squash commit together => a single commit must be better I guess.

**Comment by loicmathieu** [comment]
@dufoli it didn't incude #4529 and it has been rebased yesterday not today ...
When #4529 will be merged (soon I hope), I will have multiple open PR to rebase on it as I have multiple PRs around MongoDB at the moment ...

For the multiple commits, I know, I will do it before merge :)

File: docs/src/main/asciidoc/mongodb-panache.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/mongodb-panache.adoc]:
```suggestion
The same query facility exists for the reactive types, but the `stream()` methods act differently: they return a reactive stream `Publisher` instead of a `Stream`.
```

File: docs/src/main/asciidoc/mongodb-panache.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/mongodb-panache.adoc]:
```suggestion
It allows more advanced reactive use cases, for example you can use it to send server-sent events (SSE) via RESTEasy:
```

File: docs/src/main/asciidoc/mongodb-panache.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/mongodb-panache.adoc]:
```suggestion
TIP: `@SseElementType(MediaType.APPLICATION_JSON)` tells RESTEasy to serialize the object in JSON.
```

File: docs/src/main/asciidoc/mongodb-panache.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/mongodb-panache.adoc]:
```suggestion
TIP: If you use MongoDB with Panache in conjunction with RESTEasy, you can directly return a reactive type inside your JAX-RS resource endpoint as RESTEasy will handle it correctly.
```

**Review by gsmet — DISMISSED** [review]
I added some small suggestions for the documentation.

Otherwise, I suppose it should be merged if @FroMage and @cescoffier are OK with it. I don't really have opinions about that one.

So feel free to dismiss my review and merge as soon as my doc comments are addressed.

**Comment by loicmathieu** [comment]
@gsmet I apply all your changes. Ready to be merged ;)

**Comment by loicmathieu** [comment]
@gsmet can we merge this one ?

**Comment by gastaldi** [comment]
With Mutiny now merged in master, I think it's feasible to change this PR to use that instead of Axle?

**Comment by loicmathieu** [comment]
@gastaldi Munity is not yet integrated into the `mongodb-client` extension, when it will be done I'll integrate it.

The plan (it was more than two months ago) was to propose both implementations, I think that it will also be the case for Resteasy : you can use `CompletionStage` or Mutiny ?

Anyway, this PR is waiting since a very long time, I prefere to merge it as is and implements Munity when it will be available.

**Comment by gastaldi** [comment]
I'm fine with that, it's just that Axle is now deprecated, so I am not sure if we should keep promoting it. Can you squash this PR into a single commit please?

**Comment by loicmathieu** [comment]
@gastaldi there is one commit for the change in `mongodb-client` and one for the changes in `mongodb-panache` I tought it was better to keep one PR by extension.

**Comment by gastaldi** [comment]
Looks like it needs to be rebased, some recent commit caused that

**Comment by loicmathieu** [comment]
Yes, mines, because you merged another PR ;)
Do you want me to squashed the commits or are you OK with one per extension ?

**Comment by gastaldi** [comment]
> Yes, mines, because you merged another PR ;)

Oops, sorry about that :)

> Do you want me to squashed the commits or are you OK with one per extension ?

I'm ok with one per extension

**Comment by loicmathieu** [comment]
@gastaldi rebase done, let's wait for the CI ...

**Comment by loicmathieu** [comment]
@gastaldi it fails on an unrelated test (artemis JMS), can you relaunch the test ?

**Comment by gastaldi** [comment]
Done, let's see what happens

**Comment by loicmathieu** [comment]
@gastaldi it's green !!!

**Comment by gastaldi** [comment]
Thanks! Merged

**Pull request merged**
